storage.go

```diff
@@ -2,35 +2,192 @@ package hdfs
-	panic("not implemented")
+	rp := s.getAbsPath(path)
+	err = s.hdfs.Remove(rp)
+	if err != nil && errors.Is(err, os.ErrNotExist) {
+		// Omit `file not exist` error here
+		// ref: [GSP-46](https://github.com/beyondstorage/specs/blob/master/rfcs/46-idempotent-delete.md)
+		err = nil
+	}
+	if err != nil {
+		return err
+	}
+	return nil
 }
 
 func (s *Storage) list(ctx context.Context, path string, opt pairStorageList) (oi *ObjectIterator, err error) {
-	panic("not implemented")
+	rp := s.getAbsPath(path)
+	if !opt.HasListMode || opt.ListMode.IsDir() {
+		nextFn := func(ctx context.Context, page *ObjectPage) error {
+			dir, err := s.hdfs.ReadDir(rp)
+			if err != nil {
+				return err
+			}
+			for _, f := range dir {
+				o := NewObject(s, true)
+				o.Path = s.getFileName(f.Name())
+				if f.IsDir() {
+					o.Mode |= ModeDir
+				} else {
+					o.Mode |= ModeRead
+				}
+
+				o.SetContentLength(int64(f.Size()))
+				page.Data = append(page.Data, o)
+			}
+			return IterateDone
+		}
+		oi = NewObjectIterator(ctx, nextFn, nil)
+		return
+	} else {
+		return nil, services.ListModeInvalidError{Actual: opt.ListMode}
+	}
 }
 
 func (s *Storage) metadata(opt pairStorageMetadata) (meta *StorageMeta) {
-	panic("not implemented")
+	meta = NewStorageMeta()
+	meta.WorkDir = s.workDir
+	return meta
 }
 
 func (s *Storage) read(ctx context.Context, path string, w io.Writer, opt pairStorageRead) (n int64, err error) {
-	panic("not implemented")
+	rp := s.getAbsPath(path)
+	file, err := s.hdfs.Open(rp)
+	if err != nil {
+		return 0, err
+	}
+	if opt.HasOffset {
+		_, err = file.Seek(opt.Offset, 0)
+		if err != nil {
+			return 0, err
+		}
+	}
+
+	var rc io.Reader
+	rc = file
+
+	if opt.HasIoCallback {
+		rc = iowrap.CallbackReader(rc, opt.IoCallback)
+	}
+
+	return io.Copy(w, rc)
 }
 
 func (s *Storage) stat(ctx context.Context, path string, opt pairStorageStat) (o *Object, err error) {
-	panic("not implemented")
+
+	rp := s.getAbsPath(path)
+
+	stat, err := s.hdfs.Stat(rp)
+	if err != nil {
+		return nil, err
+	}
+
+	o = s.newObject(true)
+	o.ID = rp
+	o.Path = path
+
+	if stat.IsDir() {
+		o.Mode |= ModeDir
+		return
+	}
+
+	if opt.HasObjectMode && opt.ObjectMode.IsDir() {
+		o.Mode |= ModeDir
+	} else {
+		o.Mode |= ModeRead
+	}
+
+	o.SetContentLength(*Size)
+
+	var sm ObjectSystemMetadata
+
+	sm.Name = stat.Name()
+	sm.Size = stat.Size()
+	sm.Mode = (uint32)(stat.Mode())
+	sm.ModTime = stat.ModTime()
+	sm.IsDir = stat.IsDir()
+	sm.Sys = stat.Sys()
+
+	o.SetSystemMetadata(sm)
+
+	return o, nil
 }
 
 func (s *Storage) write(ctx context.Context, path string, r io.Reader, size int64, opt pairStorageWrite) (n int64, err error) {
-	panic("not implemented")
+	rp := s.getAbsPath(path)
+
+	filename := s.getFileName(path)
+
+	dir := path2.Dir(rp)
+	err = s.hdfs.MkdirAll(dir, 0666)
+	if err != nil {
+		return 0, err
+	}
+	file := dir + "/" + filename
+
+	var f io.Writer
+
+	_, err = s.hdfs.Stat(file)
+	if os.IsNotExist(err) {
```

Comment:
How about check the file and remove it if exist, and then `create`?

Reply:
Yes, this needs to be streamlined